sound/soc/sof/control.c

```diff
@@ -11,8 +11,41 @@
 /* Mixer Controls */
 
 #include <linux/pm_runtime.h>
+#include <linux/leds.h>
 #include "sof-priv.h"
 
+static void update_acpi_mute_led(struct snd_sof_control *scontrol,
+				 struct snd_kcontrol *kcontrol,
+				 struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_sof_dev *sdev = scontrol->sdev;
+	unsigned int mask;
+	unsigned int val;
+
+	mask = 1U << snd_ctl_get_ioffidx(kcontrol, &ucontrol->id);
+
+	if (ucontrol->value.integer.value[0] ||
+	    ucontrol->value.integer.value[1])
+		scontrol->led_ctl.led |= mask;
+	else
+		scontrol->led_ctl.led &= ~mask;
+
+	val = !scontrol->led_ctl.led;
+
+	if (val == scontrol->led_ctl.led_value)
+		return;
+
+	scontrol->led_ctl.led_value = val;
+
+	dev_dbg(sdev->dev, "sof: call_mute_led_update val:%u dir %u\n",
+		val, scontrol->led_ctl.direction);
+
+	if (!scontrol->led_ctl.direction)
+		ledtrig_audio_set(LED_AUDIO_MUTE, val ? LED_OFF : LED_ON);
+	else
+		ledtrig_audio_set(LED_AUDIO_MICMUTE, val ? LED_OFF : LED_ON);
+}
+
```

Comment:
don't you need to conditionally compile this? What happens if CONFIG_LEDS_TRIGGER_AUDIO is not enabled?

Reply:
Also what happens if you have multiple PGA controls with mute (eg. on playback you can have multiple independent outputs, if you mute one with alsamixer would the mute LED be on?).

What also happens if you have the headset handled by the HDaudio codec and the DMIC by the PCH? where would the mute control be located?

Reply:
I thinks linux/leds.h has "empty" functions defined if not compiled with CONFIG_LEDS_TRIGGER_AUDIO. So they just don't do anything...

Reply:
hd audio headset should be handled by hd audio quirks and dmic by this thing. So I assume they are different mutes and specified with ucm paths. I can actually test this with my current test device (but I just don't have leds). I think @lyakh mentioned that in windows side the mute state is remembered with different routings but I'm not sure how linux distros handle this

Reply:
on traditional devices, DMIC and headset are actually handled by the HDaudio codec so no such case of independent outputs.
With DMIC handled on the PCH, we have a different behavior now and by construction a different topology.

Reply:
> I thinks linux/leds.h has "empty" functions defined if not compiled with CONFIG_LEDS_TRIGGER_AUDIO. So they just don't do anything...

the HDaudio code comments out the entire section related to leds, so can you look into this please?

Reply:
Yes, I see the code. But I mean, it is just not compiling in couple of functions (save binary space?). I tried this PR without the CONFIG_LEDS_TRIGGER_AUDIO and it works, so I'm calling empty led functions.